recipes/libnice/0001-nicesrc-spin-the-agent-mainloop-in-a-separate-thread.patch

```diff
@@ -0,0 +1,206 @@
+Date: Tue, 13 Oct 2015 12:49:19 +1100
+Subject: [PATCH] nicesrc: spin the agent mainloop in a separate thread
+
+Don't run the mainloop from the srcpad task, since that can get blocked in the
+pipeline and cause unnecessary STUN retrasmissions (at best) and completely
+block the agent (at worst).
+---
+ gst/gstnicesrc.c | 118 ++++++++++++++++++++++++-------------------------------
+ gst/gstnicesrc.h |   4 +-
+ 2 files changed, 53 insertions(+), 69 deletions(-)
+
+diff --git a/gst/gstnicesrc.c b/gst/gstnicesrc.c
+index d369e09..028cf6d 100644
+--- a/gst/gstnicesrc.c
++++ b/gst/gstnicesrc.c
+@@ -48,6 +48,14 @@ GST_DEBUG_CATEGORY_STATIC (nicesrc_debug);
+ 
+ #define BUFFER_SIZE (65536)
+ 
++static gboolean
++gst_nice_src_start (
++    GstBaseSrc *basesrc);
++
++static gboolean
++gst_nice_src_stop (
++    GstBaseSrc *basesrc);
++
+ static GstFlowReturn
+ gst_nice_src_create (
+   GstPushSrc *basesrc,
+@@ -57,10 +65,6 @@ static gboolean
+ gst_nice_src_unlock (
+     GstBaseSrc *basesrc);
+ 
+-static gboolean
+-gst_nice_src_unlock_stop (
+-    GstBaseSrc *basesrc);
+-
+ static void
+ gst_nice_src_set_property (
+   GObject *object,
+@@ -116,8 +120,9 @@ gst_nice_src_class_init (GstNiceSrcClass *klass)
+   gstpushsrc_class->create = GST_DEBUG_FUNCPTR (gst_nice_src_create);
+ 
+   gstbasesrc_class = (GstBaseSrcClass *) klass;
++  gstbasesrc_class->start = GST_DEBUG_FUNCPTR (gst_nice_src_start);
++  gstbasesrc_class->stop = GST_DEBUG_FUNCPTR (gst_nice_src_stop);
+   gstbasesrc_class->unlock = GST_DEBUG_FUNCPTR (gst_nice_src_unlock);
+-  gstbasesrc_class->unlock_stop = GST_DEBUG_FUNCPTR (gst_nice_src_unlock_stop);
+ 
+   gobject_class = (GObjectClass *) klass;
+   gobject_class->set_property = gst_nice_src_set_property;
+@@ -179,9 +184,44 @@ gst_nice_src_init (GstNiceSrc *src)
+   src->component_id = 0;
+   src->mainctx = g_main_context_new ();
+   src->mainloop = g_main_loop_new (src->mainctx, FALSE);
+-  src->unlocked = FALSE;
+-  src->idle_source = NULL;
+   src->outbufs = g_queue_new ();
++  src->agent_io_thread = NULL;
++  g_cond_init (&src->outcond);
++}
++
++static gpointer
++gst_nice_src_agent_io_thread (gpointer data)
++{
++  GstNiceSrc *nicesrc = GST_NICE_SRC (data);
++
++  GST_INFO_OBJECT (nicesrc, "starting agent io thread");
++  g_main_loop_run (nicesrc->mainloop);
++  GST_INFO_OBJECT (nicesrc, "exiting agent io thread");
++}
++
++static gboolean
++gst_nice_src_start (GstBaseSrc * basesrc)
++{
++  GstNiceSrc *nicesrc = GST_NICE_SRC (basesrc);
++  gchar *thread_name;
++
++  thread_name = g_strdup_printf ("%s:agent_io", GST_OBJECT_NAME (nicesrc));
++  nicesrc->agent_io_thread = g_thread_new (thread_name, gst_nice_src_agent_io_thread, nicesrc);
++  g_free (thread_name);
++
++  return TRUE;
++}
++
++static gboolean
++gst_nice_src_stop (GstBaseSrc * basesrc)
++{
++  GstNiceSrc *nicesrc = GST_NICE_SRC (basesrc);
++
++  g_main_loop_quit (nicesrc->mainloop);
++  g_thread_join (nicesrc->agent_io_thread);
```

Comment:
This is racy btw: stop() could be called before the g_main_loop_run() happened, and then g_thread_join() will wait forever